File: Data Structures/Binary Search Tree/Implementations/Library-Based Implementaion/BST_Aided.cpp
#include <iostream>
#include <set>
using namespace std;

int main() {
    set<int> treeSet;
    int choice, x, y;

    while (true) {
        cout << "\nSet Operations Menu:\n";
        cout << "1. Insert Element\n";
        cout << "2. Display Elements (In-order Traversal)\n";
        cout << "3. Search for an Element\n";
        cout << "4. Remove an Element\n";
        cout << "5. Range Query (Lower and Upper Bound)\n";
        cout << "6. Exit\n";
        cout << "Enter your choice: ";
        cin >> choice;

        switch (choice) {
        case 1:
            cout << "Enter element to insert: ";
            cin >> x;
            treeSet.insert(x);
            cout << "Element inserted.\n";
            break;

        case 2:
            cout << "Sorted Elements: ";
            for (int elem : treeSet) {
                cout << elem << " ";
            }
            cout << endl;
            break;

        case 3:
            cout << "Enter element to search: ";
            cin >> x;
            if (treeSet.find(x) != treeSet.end()) {
                cout << "Element found.\n";
            } else {
                cout << "Element not found.\n";
            }
            break;

        case 4:
            cout << "Enter element to remove: ";
            cin >> x;
            if (treeSet.erase(x)) {
                cout << "Element removed.\n";
            } else {
                cout << "Element not found in the set.\n";
            }
            break;

        case 5:
            cout << "Enter lower bound: ";
            cin >> x;
            cout << "Enter upper bound: ";
            cin >> y;
            cout << "Elements in range [" << x << ", " << y << "]: ";
            for (auto it = treeSet.lower_bound(x); it != treeSet.upper_bound(y); ++it) {
                cout << *it << " ";
            }
            cout << endl;
            break;

        case 6:
            cout << "Exiting program.\n";
            return 0;

        default:
            cout << "Invalid choice. Please try again.\n";
        }
    }

    return 0;
}

File: Data Structures/Binary Search Tree/Implementations/Library-Based Implementaion/BST_Aided.java
import java.util.Scanner;
import java.util.TreeSet;

public class BST_Aided {
    public static void main(String[] args) {
        TreeSet<Integer> bst = new TreeSet<>();
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("\nTreeSet Operations:");
            System.out.println("1. Insert Element");
            System.out.println("2. Display Elements (In-order Traversal)");
            System.out.println("3. Search for an Element");
            System.out.println("4. Perform Range Query");
            System.out.println("5. Remove an Element");
            System.out.println("6. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();

            switch (choice) {
                case 1 -> {
                    System.out.print("Enter element to insert: ");
                    int insertElement = scanner.nextInt();
                    bst.add(insertElement);
                    System.out.println("Element inserted.");
                }

                case 2 -> System.out.println("Elements in TreeSet (sorted): " + bst);

                case 3 -> {
                    System.out.print("Enter element to search: ");
                    int searchElement = scanner.nextInt();
                    if (bst.contains(searchElement)) {
                        System.out.println("Element found in TreeSet.");
                    } else {
                        System.out.println("Element not found in TreeSet.");
                    }
                }

                case 4 -> {
                    System.out.print("Enter lower bound (inclusive): ");
                    int lowerBound = scanner.nextInt();
                    System.out.print("Enter upper bound (exclusive): ");
                    int upperBound = scanner.nextInt();
                    System.out.println("Elements in range [" + lowerBound + ", " + upperBound + "): " + bst.subSet(lowerBound, upperBound));
                }

                case 5 -> {
                    System.out.print("Enter element to remove: ");
                    int removeElement = scanner.nextInt();
                    if (bst.remove(removeElement)) {
                        System.out.println("Element removed.");
                    } else {
                        System.out.println("Element not found in TreeSet.");
                    }
                }

                case 6 -> {
                    System.out.println("Exiting program.");
                    scanner.close();
                    return;
                }

                default -> System.out.println("Invalid choice. Please try again.");
            }
        }
    }
}
